---
title: "Deciphering the immune dynamics according to homologous recombination status
subtitle: "Master's thesis program" 
author: "Juan A. Marín Jiménez"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd("/Users/juana.marinjimenez/Documents/TFM")
```

# R update and Seurat package installation

I followed installation instructions for Seurat_v5 at https://satijalab.org/seurat/articles/install_v5.html but needed to reinstall a 
lot of dependencies before

```{r}
if(0){
  # R update needed
  # Save packages
  tmp <- installed.packages()
  installedpkgs <- as.vector(tmp[is.na(tmp[,"Priority"]), 1])
  save(installedpkgs, file="installed_old.rda")
  # Install R here
  # Recover packages
  load("installed_old.rda")
  tmp <- installed.packages()
  installedpkgs.new <- as.vector(tmp[is.na(tmp[,"Priority"]), 1])
  missing <- setdiff(installedpkgs, installedpkgs.new)
  install.packages(missing)
  update.packages()
  
  # Seurat installation
  install.packages('Seurat')
  
  # Other recommended packages
  setRepositories(ind = 1:3, addURLs = c('https://satijalab.r-universe.dev', 'https://bnprks.r-universe.dev/'))
  install.packages(c("BPCells", "presto", "glmGamPoi"))
  
  # Install the remotes package
  if (!requireNamespace("remotes", quietly = TRUE)) {
    install.packages("remotes")
  }
  install.packages('Signac')
  remotes::install_github("satijalab/seurat-data", quiet = TRUE)
  remotes::install_github("satijalab/azimuth", quiet = TRUE)
  remotes::install_github("satijalab/seurat-wrappers", quiet = TRUE)
}
```

# Needed libraries

```{r}
if(1){
  library(Seurat)
  library(tidyverse)
  library(ggplot2)
  library(patchwork)
}
```

# GSE dataset download

Due to lack of memory, I will manually download all the files

```{r}
if(0){
  # Install GEOquery
  if (!requireNamespace("GEOquery", quietly = TRUE)) {
    install.packages("GEOquery")
  }
  library(GEOquery)
  # Dataset: Zhang_2022
  # gse_id <- "GSE165897"
  # gse_data <- getGEO(gse_id, GSEMatrix = TRUE)
  # gse_info <- getGEOSuppFiles(gse_id)
}
```

# Seurat object creation

## GSE165897 - Zhang, 2022

```{r}
if(0){
  # metadata
  cell_info <- data.table::fread("GSE165897/GSE165897_cellInfo_HGSOC.tsv.gz") %>% 
    column_to_rownames(var = 'cell')
  
  # exploring cell metadata and annotation categories
  cbind(table(cell_info$cell_subtype))
  table(cell_info$cell_subtype, cell_info$cell_type)
  
  # count matrix
  umi_counts <- data.table::fread("GSE165897/GSE165897_UMIcounts_HGSOC.tsv.gz") %>% 
    column_to_rownames(var = 'V1')
  # umi_counts[1:3,1:3]
  # umi_counts_HOLD <- umi_counts
  
  # Create my Seurat object
  # seu_zhang_22 <- CreateSeuratObject(counts=umi_counts, meta.data=cell_info)
  # mem.maxVSize()
  # Imposible to create seurat object due to vector memory limit of 16Gb
  
  # I split the matrix in 5000 colums fragments and put them in a list
  seu_zhang_22 <- list()
  for (i in 1:11){
    a1 <- i*5000-4999
    a2 <- i*5000
    if(i==11){
      a2 <- ncol(umi_counts)
    }
    umi_frag <- umi_counts[,a1:a2]
    cell_frag<- cell_info[colnames(umi_frag),]
    print(all(rownames(cell_frag)==colnames(umi_frag)))
    tmp <- CreateSeuratObject(counts=umi_frag, meta.data=cell_frag)
    seu_zhang_22[[i]] <- tmp
    tmp <- NULL
  }

  # Merging all the elements from seu_zhang_22
  seurat_zhang_22 <- seu_zhang_22[[1]]
  for (i in 2:length(seu_zhang_22)){
    seurat_zhang_22 <- merge(seurat_zhang_22, seu_zhang_22[[i]])
  }
  seurat_zhang_22
  Layers(seurat_zhang_22)
  
  # Need to fuse all the generated layers using JoinLayers
  seurat_zhang_22[["RNA"]] <- JoinLayers(seurat_zhang_22[["RNA"]])
  seurat_zhang_22
  Layers(seurat_zhang_22)
  
  # Save seurat object
  #saveRDS(seurat_zhang_22, file="./Seurat/GSE165897_seurat_raw")
}
```
## GSE147082 - Olalekan, 2021

```{r}
archive_extract("GSE147082/GSE147082_RAW.tar", dir = "GSE147082/")

data_path <- "GSE147082/"
files <- list.files(path = data_path, pattern = "^GSM.*\\.csv\\.gz$", full.names = TRUE)
seu_olalekan_21 <- list()
for (file in files) {
  data <- read.csv(file, row.names = 1)
  seurat_obj <- CreateSeuratObject(counts = as.matrix(data))
  seu_olalekan_21[[file]] <- seurat_obj
}

# Merging all the seurat objects
seurat_olalekan_21 <- merge(
  x = seu_olalekan_21[[1]], 
  y = seu_olalekan_21[2:5], # Exclude GSM4416539 due to Müllerian histology
  add.cell.ids = 1:5
)

seurat_olalekan_21@meta.data<-seurat_olalekan_21@meta.data %>% 
  mutate(orig.ident=gsub("_[A-Z]*","",rownames(.)))

# seurat_olalekan_21 <- Reduce(function(x, y) merge(x, y), seu_olalekan_21)
seurat_olalekan_21[["RNA"]] <- JoinLayers(seurat_olalekan_21[["RNA"]])
Layers(seurat_olalekan_21)

# saveRDS(seurat_olalekan_21, file = "Seurat/GSE147082_seurat_raw.rds")

```

# Seurat object analysis

The objective is to develop an script to get a QC report of every Seurat object
created for the project

Parameters:
- nCount_RNA
- nFeature_RNA
- perc.mt
- perc.ribo
- perc.hem
- perc.plat

```{r}
if(0){
  gc()
  seurat <- readRDS("Seurat/GSE165897_seurat_raw.rds")
  head(seurat@meta.data, 10)
  colnames(seurat@meta.data)
  
  # mitochondrial read content per cell
  seurat[["perc.mt"]] <- PercentageFeatureSet(seurat, pattern = "^MT-")
  # ribosomal read content per cell
  seurat[["perc.ribo"]] <- PercentageFeatureSet(seurat, pattern = "RB[SL]")
  # hemoglobin genes - includes all genes starting with HB except HBP.
  seurat[["perc.hem"]] <- PercentageFeatureSet(seurat, pattern = "^HB[^(P|E|S)]")
  # platelet markers
  seurat[["perc.plat"]] <- PercentageFeatureSet(seurat, pattern = "PECAM1|PF4")
  
  # ploting QC metrics as violin plots
  qc_features <- c("nCount_RNA", "nFeature_RNA", "perc.mt", "perc.ribo", "perc.hem", "perc.plat")
  # Complete dataset
  VlnPlot(seurat, features = qc_features, ncol = 6)
  # By samples
  VlnPlot(seurat, 
          group.by = "patient_id", 
          split.by = "treatment_phase", 
          features = "perc.mt", 
          pt.size = 0.1)
  
  # FeatureScatter to visualize feature-feature relationships
  plot1 <- FeatureScatter(seurat, feature1 = "nCount_RNA", feature2 = "perc.mt")
  plot2 <- FeatureScatter(seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
  CombinePlots(plots = list(plot1, plot2))
}
```
# Filter cells

Need to consensuate criteria. I propose:
1) Apply this to CreateSeuratObject()
min.cells = 10 # Include features detected in at least 10 cells.
min.features = 100 # Include cells where at least 100 features are detected
2) Subset the seurat:
seurat_object <- subset(seurat_object, subset = nFeature_RNA > 100) [...]

Filter by mitocondrial content - it seems like some filtering has been already applied

```{r}
if(0){
  #filters
  saveRDS(seurat, file="./Seurat/GSE165897_seurat_filtered.rds")
}
```

# Dataset individual processing

Satija's lab pipeline

```{r}
if(0){
  seurat <- readRDS("./Seurat/GSE165897_seurat_filtered.rds")
FD# Normalize
  seurat <- NormalizeData(seurat, normalization.method = "LogNormalize", scale.factor = 10000)
  # Find most variable genes
  seurat <- FindVariableFeatures(seurat)
  # Scale regressing the mitochondrial content
  seurat <- ScaleData(seurat, vars.to.regress = "perc.mt")
  # Run PCA
  seurat <- RunPCA(seurat, npcs = 100, ndims.print = 1:5, nfeatures.print = 5)
  ElbowPlot(seurat, ndims = 100)
  # Selecting the 30 first PC for tSNE and UMAP clustering
  ncomp<-30
  seurat<-FindNeighbors(seurat, reduction = "pca", dims = 1:ncomp, nn.eps = 0.5)
  # kNN graph and clustering at several resolutions
  for(i in c(seq(0.1,1,0.1), 1.5, 2, 2.5, 3, 5, 10)) {
    seurat<-FindClusters(seurat, resolution = i, n.start=10)
  }
  # seurat<-RunTSNE(seurat, dims= 1:ncomp, tsne.method = "Rtsne", nthreads=16, 
  # max_iter=2000, check_duplicates = FALSE)
  seurat <- RunUMAP(seurat, dims = 1:ncomp, min.dist = 0.75)
  # saveRDS(seurat, file = "Seurat/GSE165897_seurat_norm.rds")
}
```
# UMAP

```{r}
seurat <- readRDS("Seurat/GSE165897_seurat_norm.rds")
seurat@meta.data
p1 <- DimPlot(seurat, reduction = "umap", 
              group.by = "patient_id", 
              label = T, label.box = T) + theme(aspect.ratio=1)
p2 <- DimPlot(seurat, reduction = "umap", 
              group.by = "treatment_phase", 
              label = T, label.box = T) + theme(aspect.ratio=1)
p1|p2
```

# Data annotation

Try automated annotation using Azimuth from satija's lab

```{r annotation, fig.width=10, fig.height=8}
# options(timeout = 1200)
# BiocManager::install(c("BSgenome.Hsapiens.UCSC.hg38", "EnsDb.Hsapiens.v86", "TFBSTools"))
# devtools::install_github("satijalab/azimuth", dependencies = TRUE)
library(Azimuth)
seurat <- RunAzimuth(seurat, reference = "pbmcref")

# if(0){
#   length_subset<-50
#   cells_no<-nrow(seurat@meta.data)
#   subset<-list()
#   x <- round(cells_no/length_subset)
#   remaining<-1:cells_no
#   order<-c()
#   for (i in 1:length_subset) {
#     if(i != length_subset) {
#       subset[[i]]<-sample(remaining, x, replace=F)
#       remaining<-remaining[-which(remaining %in% subset[[i]])]
#       order<-c(order, subset[[i]])
#     }
#     if(i == length_subset) {
#       subset[[i]]<-remaining
#       order<-c(order, subset[[i]])
#     }
#   }
#   seurat$Azimuth_CT_1<-NA
#   seurat$Azimuth_CT_2<-NA
#   seurat$Azimuth_CT_3<-NA
#   
#   for (i in 1:length(subset)){
#     seurat_tmp<-subset(seurat, cells=names(seurat$orig.ident)[subset[[i]]])
#     seurat_tmp<-RunAzimuth(seurat_tmp, reference = "pbmcref")
#     
#     p <- intersect(names(seurat_tmp$orig.ident), names(seurat$orig.ident))
#     seurat$Azimuth_CT_1[match(p,names(seurat$orig.ident))] <- seurat_tmp$predicted.celltype.l1[match(p,names(seurat_tmp$orig.ident))]
#     seurat$Azimuth_CT_2[match(p,names(seurat$orig.ident))] <- seurat_tmp$predicted.celltype.l2[match(p,names(seurat_tmp$orig.ident))]
#     seurat$Azimuth_CT_3[match(p,names(seurat$orig.ident))] <- seurat_tmp$predicted.celltype.l3[match(p,names(seurat_tmp$orig.ident))]
#     rm(seurat_tmp);invisible(gc())
#     print(i)
#   }
# }

DimPlot(seurat, reduction = "umap", group.by = "predicted.celltype.l1", label = T, label.box = T) + theme(aspect.ratio = 1)

```
# Dataset integration

```{r}
table(seurat$patient_id)
table(seurat$anatomical_location, seurat$treatment_phase)
if (0) {
  seurat.list <- SplitObject(seurat, split.by = "patient_id")
  rm(seurat); invisible(gc())
  seurat.list <- lapply(seurat.list, FUN = function(x) {
    x <- NormalizeData(x, verbose = FALSE)
    x <- FindVariableFeatures(x, verbose = FALSE)
  })
  features <- SelectIntegrationFeatures(object.list = seurat.list, nfeatures = 1000)
  seurat.list <-lapply(seurat.list, FUN = function(x) {
    x <- ScaleData(x, features = features, verbose = FALSE)
    x <- RunPCA(x, features = features, verbose = FALSE)
  })
  anchors <- FindIntegrationAnchors(object.list = seurat.list, reduction = "rpca", dims = 1:30, anchor.features = features)
  rm(seurat); invisible(gc())
  seurat_integrated <- IntegrateData(anchorset = anchors, dim = 1:30, k.weight = 50)
  rm(anchors); invisible(gc())
  DefaultAssay(seurat_integrated) <- "integrated"
  seurat_integrated <- ScaleData(seurat_integrated, verbose = FALSE)
  seurat_integrated <- RunPCA(seurat_integrated, npcs=100, ndims.print = 1:5, nfeatures.print = 5)
  seurat_integrated <- FindNeighbors(seurat_integrated, reduction = "pca", dims = 1:75, nn.eps = 0.5)
  for(i in c(seq(0.1,1,0.1), 1.5, 2, 2.5, 3, 5, 10)) {
    seurat_integrated<-FindClusters(seurat, resolution = i, n.start=10)}
  # seurat_integrated <- RunTSNE(seurat_integrated, dims = 1:75, tsne.method = "Rtsne", nthreads = 16, max_iter = 2000, check_duplicates = FALSE)
  seurat_integrated <- RunUMAP(seurat_integrated)
}

```

# Replicate the workflow on new datasets

```{r}
getwd()
gse_paths <- list.dirs(path = ".", recursive = TRUE, full.names = TRUE)
gse_paths <- gse_paths[grepl("^\\.\\/GSE", gse_paths)]
gse_files <- lapply(gse_paths, list.files, full.names = TRUE)

if (!require("archive")) {
  install.packages("archive")
  library(archive)
}
```

## GSE191301

```{r}


archive_extract("GSE191301/GSE191301_RAW.tar", dir = "GSE191301/")
path <- "GSE191301/GSM5743307/"
data <- Read10X(data.dir = path)

# Crear un objeto Seurat con los datos leídos
seurat_obj <- CreateSeuratObject(counts = data)

# Verificar el objeto creado
print(seurat_obj)

# Guardar el objeto Seurat (opcional)
saveRDS(seurat_obj, file = "seurat_obj_from_10x.rds")

```

## GSE222556

```{r}

path <- "GSE222556/GSE222556_RAW/"  # Reemplaza con la ruta a tu carpeta
file.list <- list.dirs(path, full.names = TRUE, recursive = FALSE)

seu.list<-list()

for (i in 1:length(file.list)){
  print(file.list[i])
  test.data <- Read10X(data.dir = file.list[i])
  # Initialize the Seurat object with the raw (non-normalized data).
  seu.list[[i]] <- CreateSeuratObject(counts = test.data)
  seu.list[[i]]@meta.data$orig.ident<-names(file.list)[i]

}

# # Merge more than two Seurat objects
# seu<-merge(x = seu.list[[1]], y = seu.list[2:length(seu.list)])
# seu[["RNA"]] <- JoinLayers(seu[["RNA"]])

# saveRDS(seu, file="")

```

